chain/network/src/routing/routing_table_view.rs

```diff
@@ -54,159 +106,126 @@ impl RoutingTableView {
     }
 
-    pub(crate) fn set_next_hops(&mut self, routing_table: Arc<routing::NextHopTable>) {
-        self.next_hops = routing_table;
+    pub(crate) fn set_next_hops(&self, routing_table: Arc<routing::NextHopTable>) {
+        self.0.lock().next_hops = routing_table;
     }
 
     pub(crate) fn reachable_peers(&self) -> usize {
         // There is an implicit assumption here that all next_hops entries are non-empty.
         // To enforce this, we would need to make NextHopTable a newtype rather than an alias,
         // and add appropriate constructors, which would filter out empty entries.
-        self.next_hops.len()
+        self.0.lock().next_hops.len()
     }
 
     pub(crate) fn find_route(
-        &mut self,
+        &self,
         clock: &time::Clock,
         target: &PeerIdOrHash,
     ) -> Result<PeerId, FindRouteError> {
+        let mut inner = self.0.lock();
         match target {
-            PeerIdOrHash::PeerId(peer_id) => self.find_route_from_peer_id(peer_id),
+            PeerIdOrHash::PeerId(peer_id) => inner.find_route_from_peer_id(peer_id),
             PeerIdOrHash::Hash(hash) => {
-                self.fetch_route_back(clock, *hash).ok_or(FindRouteError::RouteBackNotFound)
+                inner.fetch_route_back(clock, *hash).ok_or(FindRouteError::RouteBackNotFound)
             }
         }
     }
 
-    pub(crate) fn view_route(&self, peer_id: &PeerId) -> Option<&Vec<PeerId>> {
-        self.next_hops.get(peer_id)
+    pub(crate) fn view_route(&self, peer_id: &PeerId) -> Option<Vec<PeerId>> {
+        self.0.lock().next_hops.get(peer_id).cloned()
     }
 
     /// Find peer that owns this AccountId.
-    pub(crate) fn account_owner(
-        &mut self,
-        account_id: &AccountId,
-    ) -> Result<PeerId, FindRouteError> {
-        self.get_announce(account_id)
+    pub(crate) fn account_owner(&self, account_id: &AccountId) -> Result<PeerId, FindRouteError> {
+        self.0
+            .lock()
+            .get_announce(account_id)
             .map(|announce_account| announce_account.peer_id)
             .ok_or(FindRouteError::AccountNotFound)
     }
 
     /// Add (account id, peer id) to routing table.
     /// Note: There is at most one peer id per account id.
-    pub(crate) fn add_account(&mut self, announce_account: AnnounceAccount) {
+    pub(crate) fn add_account(&self, announce_account: AnnounceAccount) {
+        let mut inner = self.0.lock();
         let account_id = announce_account.account_id.clone();
-        self.account_peers.put(account_id.clone(), announce_account.clone());
-
+        inner.account_peers.put(account_id.clone(), announce_account.clone());
         // Add account to store
-        if let Err(e) = self.store.set_account_announcement(&account_id, &announce_account) {
+        if let Err(e) = inner.store.set_account_announcement(&account_id, &announce_account) {
             warn!(target: "network", "Error saving announce account to store: {:?}", e);
         }
     }
```

Comment:
Not super happy about the current state. Here we are locking a global data-structure while accessing the storage. This is fine as an intermediate step, but I wish long-term to make sure that long-term we minimize the scope of locks. This type is basically "lock whenever I do anything", which is error-prone.

Reply:
> It was already under a global mutex of PeerManagerActor, and now it gets its own global mutex

Not quite: while in PMA, we had transactional semantics: we removed *a bunch* of accounts ate the same time, in lockstep with changing next hops, etc. Now it is possible to observe intermediate states.

Reply:
ok, will fix that

Reply:
done, PTAL